### PR TITLE
Added docs + docstrings

### PR DESCRIPTION
**Pull request by kushalbakshi**

### CODE REVIEW

**Comment by kushalbakshi** [comment]
All the pages are populated with content in what should be its final version. Will be tagging up with Drew tomorrow to try to render it locally and proof-read/check links

File: docs/src/citation.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# Citation
+
+If your work uses this Element, please cite the following manuscript and Research Resource Identifier (RRID):
+
++ Yatsenko D, Nguyen T, Shen S, Gunalan K, Turner CA, Guzman R, Sasaki M, Sitonic D, Reimer J, Walker EY, Tolias AS. DataJoint Elements: Data Workflows for Neurophysiology. bioRxiv. 2021 Jan 1. doi: https://doi.org/10.1101/2021.03.30.437358
+
++ DataJoint Elements ([RRID:SCR_021894](https://scicrunch.org/resolver/SCR_021894)) - Element Array Electrophysiology (version `<Enter version number>`)
```
Comment (CBroz1) [inline docs/src/citation.md:7]: Please add [automated versioning to the citation page](https://github.com/datajoint/element-deeplabcut/commit/352c44f98325cb79a249026d84310bff8417e7e2)

File: docs/src/concepts.md [hunk 2]
```diff
@@ -0,0 +1,116 @@
+# Concepts
+
+## Acquisition Tools for Electrophysiology
+
+### Neuropixels Probes
+Neuropixels probes were developed by a collaboration between HHMI Janelia, industry partners, and others (Jun et al., Nature 2017). Since their initial release in October 2018, 300 labs have ordered 1200 probes. Since the rollout of Neuropixels 2.0 in October 2020, IMEC has been shipping 100+ probes monthly (correspondence with Tim Harris).
```
Comment (CBroz1) [inline docs/src/concepts.md:6]: 1. Please hard wrap
2. Please implement the convention present in electrode-localization for references

File: docs/src/concepts.md [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+| Table | Description |
+| --- | --- |
+| Session | Unique experimental session identifier |
+
+### `probe` schema ([API docs](../api/element_array_ephys/probe))
+Tables related to the Neuropixels probe and electrode configuration.
+
+| Table | Description |
+| --- | --- |
+| ProbeType | a lookup table specifying the type of Neuropixels probe (e.g. "neuropixels 1.0", "neuropixels 2.0 single-shank"). |
```
Comment (CBroz1) [inline docs/src/concepts.md:70]: Please follow the convention above, with the description in sentence case

File: docs/src/concepts.md [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+| --- | --- |
+| Subject | Basic information of the research subject |
+
+### `session` schema ([API docs](https://datajoint.com/docs/elements/element-session/api/element_session/session_with_datetime))
+
+| Table | Description |
+| --- | --- |
+| Session | Unique experimental session identifier |
+
+### `probe` schema ([API docs](../api/element_array_ephys/probe))
```
Comment (CBroz1) [inline docs/src/concepts.md:65]: please verify that this link works in local deployment

File: docs/src/concepts.md [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+Incorporation of SpikeInterface into the Array Electrophysiology Element will be on DataJoint Elements development roadmap. Dr. Loren Frank has led a development effort of a DataJoint pipeline with SpikeInterface framework and NeurodataWithoutBorders format integrated (https://github.com/LorenFrankLab/nwb_datajoint).
+
+## Data Export and Publishing
+
+Element Array Electrophysiology supports exporting of all data into standard Neurodata Without Borders (NWB) files. This makes it easy to share files with collaborators and publish results on [DANDI Archive](https://dandiarchive.org/). [NWB](https://www.nwb.org/), as an organization, is dedicated to standardizing data formats and maximizing interoperability across tools for neurophysiology. For more information on uploading NWB files to DANDI within the DataJoint Elements ecosystem, visit our documentation for the DANDI upload feature of [Element Array Electrophysiology](datajoint.com/docs/elements/element-array-ephys/).
+
+## Roadmap
+
+Further development of this Element is community driven. Upon user requests we will continue adding features to this Element.
```
Comment (CBroz1) [inline docs/src/concepts.md:116]: @sidhulyalkar - Can you think of an example to add here? i.e. 'features to this Element, such as ...'

**Review by dimitri-yatsenko — APPROVED** [review]

File: docs/mkdocs.yaml [hunk 6]
```diff
@@ -1,19 +1,64 @@
-# ---------------------- PROJECT SPECIFIC ---------------------------
+--- # ---------------------- PROJECT SPECIFIC ---------------------------
 
-site_name: Element Array Ephys
+site_name: DataJoint Documentation
+site_url: http://localhost/docs/element/element-array-ephys
```
Comment (CBroz1) [inline docs/mkdocs.yaml:4]:
```suggestion
site_url: http://localhost/docs/elements/element-array-ephys
```

File: docs/src/concepts.md [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+    + Probe-insertion, ephys-recordings, LFP extraction, clusterings, curations, sorted units and the associated data (e.g. spikes, waveforms, etc.).
+    
+    + Store/track/manage different curations of the spike sorting results - supporting both curated clustering and kilosort triggered clustering (i.e., `no_curation`).
+
++ Ingestion support for data acquired with SpikeGLX and OpenEphys acquisition systems. 
++ Ingestion support for spike sorting outputs from Kilosort.
++ Triggering support for workflow integrated Kilosort processing.
++ Sample data and complete test suite for quality assurance.
+
+Incorporation of SpikeInterface into the Array Electrophysiology Element will be on DataJoint Elements development roadmap. Dr. Loren Frank has led a development effort of a DataJoint pipeline with SpikeInterface framework and NeurodataWithoutBorders format integrated (https://github.com/LorenFrankLab/nwb_datajoint).
```
Comment (CBroz1) [inline docs/src/concepts.md:108]: If we say this here, it should go in roadmap below

File: docs/src/index.md [hunk 8]
```diff
@@ -0,0 +1,14 @@
+# Element Array Electrophysiology
+
+This Element features DataJoint schemas for analyzing extracellular array electrophysiology data acquired with Neuropixels probes and spike sorted using Kilosort spike sorter. Each Element is a modular pipeline for data storage and processing with corresponding database tables that can be combined with other Elements to assemble a fully functional pipeline.
+
+The Element is comprised of `probe` and `ephys` schemas. There are several `ephys` schemas including `ephys_acute`, `ephys_chronic`, and `ephys_precluster` to handle several use cases of this pipeline and workflow. Visit the [Concepts page](./concepts.md) for more information about the use cases of `ephys` schemas and an explanation of the tables. To get started with building your own data pipeline, visit the [Tutorials page](./tutorials.md).  
+
+### `ephys_acute` module
```
Comment (CBroz1) [inline docs/src/index.md:7]: I think we can add a oneliner about each - This was in the original readme. And then the concepts page should at least mention them by name. Looks like we're missing mention of `no-curation` which will be important to have, as that's on Works Free tier

File: element_array_ephys/ephys_acute.py [hunk 9]
```diff
@@ -91,21 +68,21 @@ def get_ephys_root_data_dir() -> list:
...
-    get_session_directory(session_key: dict) -> str
-        Retrieve the session directory containing the
-         recorded Neuropixels data for a given Session
-        :param session_key: a dictionary of one Session `key`
-        :return: a string for relative or full path to the session directory
+    """Retrieve the session directory with Neuropixels for the given session.
+
+    Args:
+        session_key (dict): A dictionary mapping subject to an entry in the subject table, and session_datetime corresponding to a session in the database. 
+    Returns:
```
Comment (CBroz1) [inline element_array_ephys/ephys_acute.py:75]: Does it render ok with the extra space between last arg and `returns:`? I added that extra linebreak to follow convention

**Review by CBroz1 — CHANGES_REQUESTED** [review]
Thanks for your work here! I've taken a first pass on everything except the docstrings themselves

File: docs/src/concepts.md [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+
+- Andreas Tolias Lab (Baylor College of Medicine)
+
+- BrainCoGs (Princeton Neuroscience Institute)
+
+- Brody Lab (Princeton University)
+
+## Element Architecture
+
+Each of the DataJoint Elements are a set of tables for common neuroinformatics modalities to organize, preprocess, and analyze data. Each node in the following diagram is either a table in the Element itself or a table that would be connected to the Element.
```
Comment (dimitri-yatsenko) [inline docs/src/concepts.md:48]:
```suggestion
Each of the DataJoint Elements creates a set of tables for common neuroscience data modalities to organize, preprocess, and analyze data. Each node in the following diagram is a table within the Element or a table connected to the Element.
```

File: element_array_ephys/probe.py [hunk 11]
```diff
@@ -78,15 +96,16 @@ def create_neuropixels_probe(probe_type='neuropixels 1.0 - 3A'):
...
-            :param col_spacing: (um) horrizontal spacing between sites
-            :param row_spacing: (um) vertical spacing between columns
-            :param white_spacing: (um) offset spacing
-            :param col_count: number of column per shank
-            :param shank_count: number of shank
-            :param shank_spacing: spacing between shanks
-            :return:
+            """Builds electrode layouts.
+
+            Args:
```
Comment (CBroz1) [inline element_array_ephys/probe.py:101]: Please add typing suggestions to functions def. e.g., def func(arg1: int, arg2: str = "DefaultStr")

If you take this step first in the future, you can get templated args with [autoDocstring](https://marketplace.visualstudio.com/items?itemName=njpwerner.autodocstring) e.g., `arg2 (str): _description_. Defaults to 'DefaultStr'` autopopulates

File: element_array_ephys/probe.py [hunk 12]
```diff
@@ -122,6 +141,14 @@ def build_electrodes(site_count, col_spacing, row_spacing,
 
 @schema
 class Probe(dj.Lookup):
+    """Represent a physical probe with unique ID
+
+    Attributes:
+        probe (foreign key, varchar(32) ): Unique ID for this model of the probe.
+        ProbeType (query): ProbeType entry.
```
Comment (CBroz1) [inline element_array_ephys/probe.py:148]: Can this be a `dict` key? Or does it need to be a query object? If changed, may require changes elsewhere
```suggestion
        ProbeType (dict): ProbeType entry.
```

File: element_array_ephys/ephys_precluster.py [hunk 13]
```diff
@@ -314,6 +373,15 @@ class Step(dj.Part):
 
 @schema
 class PreClusterTask(dj.Manual):
+    """Defines a pre-clusting task ready to be run.
+
+    Attributes:
+        EphysRecording (foreign key): EphysRecording primary key.
+        PreclusterParamSteps (foreign key): PreClusterParam Steps primary key.
+        precluster_output_dir (varchar(255) ): relative path to directory for storing results of pre-clustering.
+        task_mode (enum('none', 'load', 'trigger') ): Pick between no pre-clustering, load results from file, or trigger automated pre-clustering.
```
Comment (CBroz1) [inline element_array_ephys/ephys_precluster.py:382]:
```suggestion
        task_mode (enum ): `none` (no pre-clustering), `load` results from file, or `trigger` automated pre-clustering.
```

File: element_array_ephys/ephys_precluster.py [hunk 14]
```diff
@@ -695,12 +861,26 @@ def make(self, key):
...
     definition = """
     # A set of spike waveforms for units out of a given CuratedClustering
     -> CuratedClustering
     """
 
     class PeakWaveform(dj.Part):
+        """Mean waveform across spikes for a given unit. 
+
+        Attributes:
+            master (foreign key): WaveformSet primary key.
```
Comment (CBroz1) [inline element_array_ephys/ephys_precluster.py:879]: I don't think `master` will make sense to new users in the context of API docs. Please change across schema
```suggestion
            WaveformSet (foreign key): WaveformSet primary key. 
```

File: element_array_ephys/ephys_acute.py [hunk 15]
```diff
@@ -23,30 +23,14 @@
...
-                    Retrieves the root directory for all processed data to be found from or written to
-                    :return: a string for full path to the root directory for processed data
+    """Activates the `ephys` and `probe` schemas. 
+
+    Args:
+        ephys_schema_name (str): A string containing the name of the ephys schema.
+        probe_schema_name (str): A string containing the name of the probe scehma.
+        create_schema (bool): If True, schema will be created in the database.
+        create_tables (bool): If True, tables related to the schema will be created in the database.
+        linking_module (str): A string containing the module name or module containing the required dependencies to activate the schema.
```
Comment (CBroz1) [inline element_array_ephys/ephys_acute.py:33]: Please readd upstream items following [this convention](https://github.com/datajoint/element-deeplabcut/blob/main/element_deeplabcut/model.py#L44-L53)

File: element_array_ephys/ephys_acute.py [hunk 16]
```diff
@@ -504,6 +577,15 @@ class ClusterQualityLabel(dj.Lookup):
 
 @schema
 class ClusteringTask(dj.Manual):
+    """A clustering task to spike sort electrophysiology datasets.
+
+    Attributes:
+        EphysRecording (foreign key): EphysRecording primary key.
+        ClusteringParamSet (foreign key): ClusteringParamSet primary key.
+        clustering_outdir_dir (varchar (255) ): Relative path to output clustering results. 
+        task_mode (enum('load', 'trigger') ): Either loads or computes clustering.
```
Comment (CBroz1) [inline element_array_ephys/ephys_acute.py:586]: Let's just label this as (enum) and then give the `options` in the description. See comment in Probe

**Review by CBroz1 — CHANGES_REQUESTED** [review]
Reviewed docstrings for probe and acute. Please request re-review when implemented across schemas.

File: docs/src/concepts.md (hunk 4 above)
Comment (kushalbakshi) [reply docs/src/concepts.md:73]: It is working in local

File: element_array_ephys/ephys_acute.py (hunk 9 above)
Comment (kushalbakshi) [reply element_array_ephys/ephys_acute.py:75]: Good catch. It wasn't rendering correctly. I've fixed it everywhere in this element.

File: element_array_ephys/probe.py (hunk 12 above)
Comment (kushalbakshi) [reply element_array_ephys/probe.py:148]: Yeah I wasn't sure about this. I think I can use a dict. I think I've gone back and forth so I'll make sure everything is consistent

File: docs/src/concepts.md (hunk 7 above)
Comment (kushalbakshi) [reply docs/src/concepts.md]: Changed it, just pending commit.

File: docs/src/index.md (hunk 8 above)
Comment (kushalbakshi) [reply docs/src/index.md:13]: Added mention of `no-curation` and a short description of each

File: docs/src/concepts.md (hunk 2 above)
Comment (kushalbakshi) [reply docs/src/concepts.md:6]: Hopefully the wrap worked. I thought VSCode was doing that automatically. 

References will be updated in the next commit

File: docs/src/concepts.md (hunk 3 above)
Comment (kushalbakshi) [reply docs/src/concepts.md:78]: Done

File: element_array_ephys/probe.py (hunk 11 above)
Comment (kushalbakshi) [reply element_array_ephys/probe.py:114]: Done. And thank you, I will use that in the future

File: element_array_ephys/ephys_acute.py (hunk 15 above)
Comment (kushalbakshi) [reply element_array_ephys/ephys_acute.py:39]: Done

File: element_array_ephys/ephys_chronic.py [hunk 17]
```diff
@@ -551,7 +551,7 @@ class ClusteringTask(dj.Manual):
         EphysRecording (foreign key): EphysRecording primary key.
         ClusteringParamSet (foreign key): ClusteringParamSet primary key.
         clustering_outdir_dir (varchar (255) ): Relative path to output clustering results. 
-        task_mode (enum('load', 'trigger') ): Either loads or computes clustering.
+        task_mode (enum): `Trigger` and `load` either computes clustering or imports existing clustering data, respectively.
```
Comment (CBroz1) [inline element_array_ephys/ephys_chronic.py:554]:
```suggestion
        task_mode (enum): `Trigger` computes clustering or and `load` imports existing data.
```

File: element_array_ephys/ephys_no_curation.py [hunk 18]
```diff
@@ -601,7 +601,7 @@ class ClusteringTask(dj.Manual):
         EphysRecording (foreign key): EphysRecording primary key.
         ClusteringParamSet (foreign key): ClusteringParamSet primary key.
         clustering_outdir_dir (varchar (255) ): Relative path to output clustering results. 
-        task_mode (enum('load', 'trigger') ): Either loads or computes clustering.
+        task_mode (enum): `Trigger` and `load` either computes clustering or imports existing clustering data, respectively.
```
Comment (CBroz1) [inline element_array_ephys/ephys_no_curation.py:604]:
```suggestion
        task_mode (enum): `Trigger` computes clustering or and `load` imports existing data.
```

File: element_array_ephys/ephys_precluster.py [hunk 19]
```diff
@@ -683,7 +683,7 @@ class ClusteringTask(dj.Manual):
         EphysRecording (foreign key): EphysRecording primary key.
         ClusteringParamSet (foreign key): ClusteringParamSet primary key.
         clustering_outdir_dir (varchar (255) ): Relative path to output clustering results. 
-        task_mode (enum('load', 'trigger') ): Either loads or computes clustering.
+        task_mode (enum): `Trigger` and `load` either computes clustering or imports existing clustering data, respectively.
```
Comment (CBroz1) [inline element_array_ephys/ephys_precluster.py:686]:
```suggestion
        task_mode (enum): `Trigger` computes clustering or and `load` imports existing data.
```

File: docs/src/concepts.md (hunk 5 above)
Comment (sidhulyalkar) [reply docs/src/concepts.md]: Can add that we will plan to add functionality to support large (48 hour) neuropixel recordings via an overlapping segmented processing approach.

**Review by CBroz1 — APPROVED** [review]

**Comment by CBroz1** [comment]
@kushalbakshi - Let me know if you want to go over resolving merge conflicts together

**Comment by kushalbakshi** [comment]
> @kushalbakshi - Let me know if you want to go over resolving merge conflicts together

Yes, are you available tomorrow morning ~ 9:30? I can send a calendar invite

File: CHANGELOG.md [hunk 20]
```diff
@@ -54,6 +58,7 @@ Observes [Semantic Versioning](https://semver.org/spec/v2.0.0.html) standard and
 + Add - Readers for: `SpikeGLX`, `Open Ephys`, `Kilosort`
 + Add - Probe table supporting: Neuropixels probes 1.0 - 3A, 1.0 - 3B, 2.0 - SS, 2.0 - MS
 
+[0.2.1]: https://github.com/datajoint/element-array-ephys/compare/0.2.0...0.2.1
 [0.2.0]: https://github.com/datajoint/element-array-ephys/releases/tag/0.2.0
 [0.1.4]: https://github.com/datajoint/element-array-ephys/compare/0.1.3...0.1.4
 [0.1.3]: https://github.com/datajoint/element-array-ephys/compare/0.1.2...0.1.3
```
Comment (CBroz1) [inline CHANGELOG.md:64]: please change all to tags links instead of compare

**Review by CBroz1 — CHANGES_REQUESTED** [review]
Sorry it's taking me so many rounds - I think we'll be ready to merge after this

**Comment by kushalbakshi** [comment]
> Sorry it's taking me so many rounds - I think we'll be ready to merge after this

No problem at all. I'm glad we're catching these issues now

**Review by CBroz1 — APPROVED** [review]
Thanks for all your work on this @kushalbakshi !

Please push a tag and let me know if anything fails in github actions

**Pull request merged**